### PR TITLE
Rk/kwdyz11 kkl15

### PR DESCRIPTION
**Pull request by kliegl**
I would like to add two new notebooks for the kwdyz11.arrow and kkl15.arrow. They demonstrate shrinkage effects for significant and non-significant fixed effects. I will still have to add a graph for the cell means.

### CODE REVIEW

**Comment by kliegl** [comment]
Added plots of means and used parametric bootstrap to compute SCIs and comparative density plots for residual, FEs, VCs, and CPs for both notebooks.

**Review by dmbates — COMMENTED** [review]
- I like the use of the density plots and relating the results to the shortest coverage intervals. It seems you mention the coverage intervals but don't display them.
- I keep getting complaints about being unable to resolve package versions.  Can you try updating the package versions by clicking on one of the check marks besides, e.g. `using Arrow` in the first code block then clicking on the upward pointing arrow in the pop-up box?  I think it is better if you do it rather than me because I am running a development version of Julia.  I can switch back to julia-v1.6.2 if you prefer I try it.

**Comment by kliegl** [comment]
> * I like the use of the density plots and relating the results to the shortest coverage intervals. It seems you mention the coverage intervals but don't display them.
> * I keep getting complaints about being unable to resolve package versions.  Can you try updating the package versions by clicking on one of the check marks besides, e.g. `using Arrow` in the first code block then clicking on the upward pointing arrow in the pop-up box?  I think it is better if you do it rather than me because I am running a development version of Julia.  I can switch back to julia-v1.6.2 if you prefer I try it.

Questions of clarification before I d
(1) I am using julia-1.7.0-rc1
(2) I did not include updates of the Project.toml since I think this file was not part of the Branch off main.  Could this be the problem? Should I include this file with committ/ push?  Also I don't want to  push packages you don't like.

**Comment by dmbates** [comment]
> > * I like the use of the density plots and relating the results to the shortest coverage intervals. It seems you mention the coverage intervals but don't display them.
> > * I keep getting complaints about being unable to resolve package versions.  Can you try updating the package versions by clicking on one of the check marks besides, e.g. `using Arrow` in the first code block then clicking on the upward pointing arrow in the pop-up box?  I think it is better if you do it rather than me because I am running a development version of Julia.  I can switch back to julia-v1.6.2 if you prefer I try it.
> 
> Questions of clarification before I d (1) I am using julia-1.7.0-rc1 (2) I did not include updates of the Project.toml since I think this file was not part of the Branch off main. Could this be the problem? Should I include this file with committ/ push? Also I don't want to push packages you don't like.

It turns out that when I switched back to julia-v1.6.2 the problems with the versions resolved themselves so no need to take any action.  I sure hope they get around to releasing v1.7.0 soon so that we can just switch to that version and also recommend `using MKL` for anyone with an Intel processor.

**Comment by dmbates** [comment]
@kliegl  Do you want to make the change to call `DataFrame(shortestcovint(...))` or shall I?

**Comment by kliegl** [comment]
> @kliegl Do you want to make the change to call `DataFrame(shortestcovint(...))` or shall I?

@dmbates. There was a lot of copy/past from kwdyz11.jl.  I missed this cell. Go ahead.

**Comment by kliegl** [comment]
@dmbates One set of default graphs that is still missing in these vignettes relate to residual diagnostics; the other one is faceting on the random factor and including pooled, within, and conditional modes. I think with those we would be set to put up vignettes of this kind. If users learn to execute this series of plots before and after fitting an LMM, the number of published degnerate LMM would quickly converge to a very small percentage for this set of users.

**Comment by dmbates** [comment]
@kliegl  Shall we standardize on using julia-v1.7.0-rc1 then?

**Comment by kliegl** [comment]
> @kliegl Shall we standardize on using julia-v1.7.0-rc1 then?
Fine with me.

File: notebooks/kkl15.jl
Comment (dmbates) [inline notebooks/kkl15.jl]: Did you intend for this plot to have both a boxplot and a violin plot?

Would it be okay with you if I transposed the plot? I keep telling students that most plots of a numeric response by a categorical covariate should have the response on the horizontal axis.  Most times you can shrink the vertical axis without losing any information because it is just four categories.

Comment (kliegl) [reply notebooks/kkl15.jl]: Actually, my preference would be a dotplot(). I was playing with boxplot() and violinploit() and did not really like either.    I agree with the coord_flip().  Can you implement this? I don't know how to do this with AoG. (Also I did  not try it, but if there is a dotplot(), feel free to exchange it.

Comment (kliegl) [reply notebooks/kkl15.jl]: PS: I am in a meeting and cannot act myself currently.

Comment (dmbates) [reply notebooks/kkl15.jl]: Okay, I will experiment,  It may take a while. (But I'm sure it will be worth the effort.)

Comment (kliegl) [reply notebooks/kkl15.jl]: 👍

Comment (dmbates) [reply notebooks/kkl15.jl]: So I am starting to learn more about the data and only now realized that there are 147 subjects.  The `Subj` variable has 86 levels but they go from "1" to "9", lexicographically.  I have changed those to "S001" to "S147" in a modified version of the notebook for you to check.  I also added a CairoMakie boxplot with the horizontal orientation and a comparative densityplot of all the response times.  These are just suggestions.

Also, I found out that there are 24 columns in the Arrow table.  Are they all necessary or are some of them derived columns?  It appears that many of them are columns extracted from a model matrix.  Storing those as separate columns is a table is a practice I would strongly discourage.

**Comment by kliegl** [comment]
@dmbates  Here I respond to all of your comments.
+ Thanks for fixing the subject labels. I think we should keep the bad labels in the file so there is a reason to show how to transform them.
+ I think we should keep all your plots; the boxplot looks very nice. I propose to add (at least) three more (see below).
+ The colums in `kkl15.arrow` allow me to reproduce earlier analyses estimating additional fixed effects, VCs, and CPs. The full model is quite complex and I would not know how to code these models without extracting indicator variables from the model matrix. Of course, after I reproduce the analyses with Julia code,  we can get rid of the columns in the file and compute them in the script. 
+ I updated the "sister" script `kwdyz11.jl` with the following:
     + changed to analyzing log-transformed rts; may just as well do it right here
     + added residual-over-fitted, q-q, and density for standardized residual and normal. 
     + added text at various places
 + Three plots are still missing, but I already left marks where I want to add them
     +  ridge plot
     + dotplot
     + overlay of conditional modes, within-subject, and pooled of GM and experimental effects 

The overlay plot may not be as important in this case because  there is no covariate on the x-axis, but I still want to get a facet plot for a subset of subjects. An alternative would be to overlay the within-subject effects with the caterpillar of conditional modes. I have seen plots of this kind by you some time. 

I will adopt these revisions for `kkl15.jl`  Feel free to improve on the `kwdyz11.jl` for now. 

Getting there ...

File: notebooks/kkl15.jl
Comment (dmbates) [inline notebooks/kkl15.jl]: The code blocks defining `dat_subj` and `dat_cond` got reversed in the displayed form of the notebook.  That's not a problem for Pluto because it will evaluate `dat_subj` before a code block that requires `dat_subj`.  However, it may be confusing to students.

Also, there are many more observations under the `val` condition than in the other conditions.  Are you concerned about comparing these measurements as if on the same scale when the standard errors of these means are quite different?  I realize that `val` is the reference condition and thus it makes sense to have it over-sampled but that probably should be mentioned at some point.

Comment (kliegl) [reply notebooks/kkl15.jl]: Yes, actually it generated a problem for me at one point. Is there a way to force an ordering according to current appearance? 

PS:  I am not so sure about Pluto for daily work, I must say. I think it may be better to work with scripts and convert a finished script to a notebook if it is to be used for teaching.  It seems you like it for daily work?

File: notebooks/kkl15.jl
Comment (dmbates) [inline notebooks/kkl15.jl]:
```suggestion
    :CTR  => SeqDiffCoding(levels=["val", "sod", "dos", "dod"]),
    :Subj => Grouping(),
```

Also, the `begin/end` is superfluous here as there is only one expression to evaluate.

File: notebooks/kkl15.jl
Comment (dmbates) [inline notebooks/kkl15.jl]: I tend to write this as
```jl
m1 = let formula  = @formula  lrt ~ 1 + CTR + (1 + CTR | Subj)
    fit(MixedModel, formula, dat; contrasts=contr1
end
```
because the `formula` identifier is local to the `let` block.  Of course, the downside is needing to explain to students what the `let` block does.

I kind of wish there was a way of overriding the compact HTML display because I always want to have a reminder of how many levels there are in the grouping factors and that doesn't show up in the coefficient table or in the `VarCorr`.  Perhaps we should add that to the `VarCorr` display?  What do you think @palday ?

File: notebooks/kkl15.jl
Comment (kliegl) [reply notebooks/kkl15.jl]: In this script, in no other script  `:Subj => Grouping()` gave me an error "Grouping" not defined. I even copied a block, restarted Julia and Pluto, etc. After a while I gave up, because I did not need it for the 86 subjects. I thought it might have been related to cells getting out of order,  but I am not sure.  It will be interesting to see whether it works after I  check out your modification.

File: notebooks/kkl15.jl
Comment (kliegl) [reply notebooks/kkl15.jl]: No objection.

File: notebooks/kkl15.jl
Comment (dmbates) [inline notebooks/kkl15.jl]: We are relying a lot on the `issingular` result when, in fact, it is a conservative indicator.  All that it is checking is whether one of the diagonal elements of \lambda matrices is exactly zero, which gets us into the area of "exactly zero" versus "very near to zero".  Also, that is checking on the relative covariance scale whereas the PCA is on the correlation scale and they could be quite different.  So I am a bit uneasy about that.

File: notebooks/kkl15.jl
Comment (palday) [reply notebooks/kkl15.jl]: we could also have a `FitDescription` or the like which has aic, bic, loglik, nobs

but I like the idea of adding it to varcorr, just need to figure out how to make it play nice with the table.

File: notebooks/kkl15.jl
Comment (kliegl) [reply notebooks/kkl15.jl]: I think in one of the other tutorial scripts, probably "MMT_basics, I describe this ambiguity and end with  "Embrace uncertainty!"

Comment (kliegl) [reply notebooks/kkl15.jl]: But I also think we should insert your comment in the script. That's what it is about.

File: notebooks/kkl15.jl
Comment (dmbates) [inline notebooks/kkl15.jl]: This says that the covariance matrix is singular but the `issingular` check above returned `false`.  In face, the condition number of \lambda is 12.2, which is quite good.  A condition number is between 1, indicating that all the singular values are equal, and infinity, indicating that the matrix is singular with one or more of the singular values being zero.  So 12 means that the ratio of the largest singular value to the smallest, which is the ratio of the axis lengths of the ellipsoidal covariance shape,  is 12, which is not bad for three dimensions.

File: notebooks/kkl15.jl
Comment (dmbates) [inline notebooks/kkl15.jl]: Do we need to switch to `StatsPlots` here?  I think it can be written as
```
data((; fitted=fitted(m1), resid=residuals(m1)) * mapping(:fitted=>"Fitted values for m1", :resid="Residuals") |> draw
```
(The default `visual` is `Scatter`.)

File: notebooks/kkl15.jl
Comment (dmbates) [inline notebooks/kkl15.jl]: These `qq` plots are confusing to me.  In my world the horizontal axis in a `qqnorm` plot should be standard normal quantiles, so the axis should be from about -3 to 3.  I think @palday had a PR changing this accepted by the AlgebraOfGraphics maintainers.

File: notebooks/kkl15.jl
Comment (kliegl) [reply notebooks/kkl15.jl]: No objection.

File: notebooks/kkl15.jl
Comment (kliegl) [reply notebooks/kkl15.jl]: This comes from StatsPlots. Is there an AoG q-q-plot?

File: notebooks/kkl15.jl
Comment (dmbates) [inline notebooks/kkl15.jl]: @palday Should we write a method so that this can be collapsed to 
```jl
caterpillar!(Figure(; resolution=(800,1000)), m1; orderby=2)
```
I don't see any pressing need to evaluate `ranefinfo` separately.

File: notebooks/kkl15.jl
Comment (dmbates) [inline notebooks/kkl15.jl]: That plot is pretty much the definition of an insignificant main effect for the attraction effect.

The `begin/end` is superfluous here because a `let` block is a single expression.  It is a matter of taste on whether to enclose what could become multiple expressions in a `begin/end` block even if they don't need it right now.

I am confused about how to interpret this.  The `attraction effect` in the sequential difference coding is not significant in the fixed effects but has a larger standard deviation than does the `object effect` in the random effects.  Also, all the correlations associated with the object effect are poorly defined.  The coverage intervals extend from roughly -1 to +0.5.  Is this related to using the sequential diff contrasts?

Comment (kliegl) [reply notebooks/kkl15.jl]: It is my favorite insignificant main effect, because of the highly reliable individual differences associated with its VC and its correlation with the spatial effect. That was the main reason I checked on the CP with the follow-up study.

File: notebooks/kkl15.jl
Comment (palday) [reply notebooks/kkl15.jl]: yeah, any plot type in Makie will work. But also we define a method in MixedModelsMakie so you can just do `qqnorm(m1)` when [Cairo]Makie is loaded.

File: notebooks/kkl15.jl
Comment (kliegl) [reply notebooks/kkl15.jl]: Another funny thing is that the insignificance is not quite fair because the physical distance was 1.44 the comparison distance (diagonal vs. horiz/vert). So one could argue that when distance is taken into account there is a strong attraction effect.

File: notebooks/kkl15.jl
Comment (palday) [reply notebooks/kkl15.jl]: yep, I would just do

```julia
caterpillar!(fig, m::MixedModel, args...; kwargs...) = caterpillar!(fig, ranefinfo(m), args...; kwargs...)
```

File: notebooks/kkl15.jl
Comment (kliegl) [reply notebooks/kkl15.jl]: Another aspect I need to add (and put here to think of it) is that the original study was carried out at Peking Normal University and the follow-up study at the University of Potsdam. Depending on where you stand with respect to cultural diversity, this was more a conceptual than  a direct replication.

File: notebooks/kkl15.jl
Comment (dmbates) [reply notebooks/kkl15.jl]: Using Pluto for development is tricky.  Ir may be easier to work on a script for developing code but then text becomes awkward.  Because of the UUID associated with each block it is not easy to go back and forth between editing in, say, VS Code, and in a Pluto notebook.  You can't just add a Markdown chunk or a code chunk in an editor and have it appear in the right place (or at all) in the notebook.

And much as I like having the Project.toml and Manifest.toml available in the Pluto notebook for when students download them, it is difficult when multiple people are developing the same notebook.

Similarly, there are good aspects about Books.jl but also a lot of tricky bits and I am just starting to learn them.

If you were to edit code in an editor like VS Code what would you do about the text chunks?

File: notebooks/kkl15.jl
Comment (dmbates) [reply notebooks/kkl15.jl]: At a time like that you can open a temporary code window in Pluto and start experimenting to see if the Grouping function is defined in the expected place.  That is, does
```jl
MixedModels.Grouping
```
exist, and, if so, what is
```jl
julia> typeof(MixedModels.Grouping)
DataType
```

File: notebooks/kkl15.jl
Comment (dmbates) [reply notebooks/kkl15.jl]: Thanks, @palday.  Let me know if you want me to look at it while you deal with crunch time.  It is not a high priority item.

File: notebooks/kkl15.jl
Comment (kliegl) [reply notebooks/kkl15.jl]: I am using 
#=
Text
=#

This can be copied quickly into a PlutoNB cell.

File: notebooks/kkl15.jl
Comment (kliegl) [reply notebooks/kkl15.jl]: Argh,  I was looking for `StatsModels.Grouping()`

File: notebooks/kkl15.jl
Comment (kliegl) [reply notebooks/kkl15.jl]: See https://github.com/JuliaMixedModels/McNeishBauerSim/pull/6/commits/36ebcb258c16bc8bebf0651d64d0eb8478efe152

Comment (kliegl) [reply notebooks/kkl15.jl]: And one  more argument: The order information at the end of the script makes the file really ugly. 

@dmbates @palday My proposal is to stay with PlutoNBs in RePsychLing/SMLP2021/notebooks -- I would soon like to send a note on the zoomApp when they are  in agreeable shape. As I said, I do think they are ok for teaching.  For the repos in JuliaMIxedModels (EmbraceUncertainty and McNeishBowerSim) we are don't need to use PlutoNBs, right?

**Comment by kliegl** [comment]
>  I realize that val is the reference condition and thus it makes sense to have it over-sampled but that probably should be mentioned at some point.

I overlooked this comment. I will add some text at the beginning -- one of the situations where violation of the balanced design is motivated by the theoretical question and calls for an LMM-.

 The imbalance holds only at the trial level; it is not "visible" in ANOVAs, but, of course, will show up in the conditional modes -- less shrinkage for spatial effect. I may also add a few sentences on this in the results.

**Comment by dmbates** [comment]
I'm getting a Positive Definite exception on the regression on age by sex of each test in `MixedModelsTutorial_basics.jl` - I think because you have changed the levels of `Sex` and are checking for specific values in the filter. Wouldn't it be better to write that section as a `combine(groupby(...))` rather than creating a DataFrame then looping to fill in values?  If I write it as
```jl
lms1 = combine(
	groupby(dat, [:Test, :Sex]),
	[:age, :zScore] => ((x,y) -> simplelinreg(x, y)) => :coef,
)
```
I get only results for `Boys`.  It appears that somewhere along the line the sample for the girls is being lost.

It also brings up the issue of "is this really basics for MixedModels?" for that notebook.  There is a lot of data manipulation and re-labelling, etc., which is relevant for a final publication but may not be where you want to start on a "basics" notebook.

I'll go look at the other notebooks now.

**Comment by dmbates** [comment]
I like the way that `kkl15.jl` flow now and the use of many plots.

**Comment by kliegl** [comment]
I am pretty sure that  corrected version of this notebook is in the other branch  rk/smlp2021_cleanup

I don't think I know how to keep branches snychronized with each other. Is there a simple solution? Or should I always work only with one branch?  

Sorry, if | screwed up here.

Added: @dmbates 
Yes, I checked.  The code in the other branch implement

```
combine(groupby(df,[:Sex, :Test]), [:age, :zScore] => simplelinreg => :coef)
```

and

```
combine(groupby(dat,[:Sex, :Test]), [:age, :zScore] => simplelinreg => :coef)
```

This is an old version and this may not be the only change. How can I get rid of this old notebook in this branch? The revision is here:
https://github.com/RePsychLing/SMLP2021/pull/57/files

**Comment by kliegl** [comment]
> It also brings up the issue of "is this really basics for MixedModels?" for that notebook. There is a lot of data manipulation and re-labelling, etc., which is relevant for a final publication but may not be where you want to start on a "basics" notebook.

@dmbates  I am not at all wedded the the name. I think I called it basics because I used it for my opening presentation. I think it could end up as a vignette of the same kind as  kkl15.jl and kwdyz11.jl.  fggk21.jl is not such a good name because I use for the script going with the publication. Perhaps we could introduce prefix for scripts that illustrate MixdModels.jl with reference to a specific publication. For examples. using `mmt`as an abbreviation for mixed model tutorial, we could name such scripts:   mmt_fggk21.j,  mmt_kkl15.jl, mmt_kwdyz11.jl,  mmt_sleepstudy.jl,  etc.

**Comment by dmbates** [comment]
Regarding getting rid of an out-of-date copy of a notebook in this branch - this is where git sometimes frustrates me.  I find it difficult to undo mistakes.

It may be best to let @palday weigh in here when he has time.  The way I would do it is a hack - I would check out a branch that has the version of the notebook to which I want to revert, keep a copy of that version in a temporary location, check out this branch and copy in the version of the notebook that I want.  This ends up with a messy git history because it has a change followed by the opposite change but I don't know enough git to do it more cleanly.

It seems that the most recent changes to that file were in commit 69e3687 a week ago Thursday so you could try looking at that.  As I said, I don't know how to undo specific changes in a commit.

Regarding the naming, it would be fine with me to use a prefix.

**Comment by kliegl** [comment]
> Regarding getting rid of an out-of-date copy of a notebook in this branch - this is where git sometimes frustrates me. I find it difficult to undo mistakes.
> 
> It may be best to let @palday weigh in here when he has time. The way I would do it is a hack - I would check out a branch that has the version of the notebook to which I want to revert, keep a copy of that version in a temporary location, check out this branch and copy in the version of the notebook that I want. This ends up with a messy git history because it has a change followed by the opposite change but I don't know enough git to do it more cleanly.
> 
> It seems that the most recent changes to that file were in commit 69e3687 a week ago Thursday so you could try looking at that. As I said, I don't know how to undo specific changes in a commit.
> 
> Regarding the naming, it would be fine with me to use a prefix.

Is it not possible for you, assuming the scripts are ok,
(1) accept the notenbook mmt_basics,jl in this pull request:
https://github.com/RePsychLing/SMLP2021/pull/57/files

Perhaps rename it to mmt_fggk21.jl before the merge.  Then I can delete the branch "rk/sml2021cleanup"

(2) accept the notebook kwdyz11.jl  (renamed to mmt_kwdyz11.jl) and kkl15.jl (renomed to mmt_kkl15.jl) and merge them. Then I can delete the branch "rk/kwdy11_kk15". 

I will work only with one branch per repo for now.  I really don't want to get into the hacking you describe.  Sorry.

**Comment by kliegl** [comment]
@dmbates @palday  The three notebooks I had asked to be reviewed are tin this branch now.
+ mmt_kwdyz11.jl 
+ mmt_kkl15.jl
+ mmt_fggk21.jl

I also renamed the other scripts that are already in the repo accordingly. So there is now  mmt_kb07.jl -- I am sure you are happy to know. 😊

**Review by dmbates — APPROVED** [review]

**Pull request merged**
